stdlib/asm/mem.masm

```diff
@@ -0,0 +1,41 @@
+# ===== MEMORY FUNCTIONS ==========================================================================
+
+#! memcopy(read_ptr: address, write_ptr: address, words: felt)
+#!
+#! Copies `words` words from `read_ptr` to `write_ptr`.
+#!
+#! Stack transition looks as follows:
+#! [read_ptr, write_ptr, words, ...] -> [...]
+export.memcopy
+  # add counter
+  # stack: [read_ptr, write_ptr, count, words, ...]
+  push.0 movdn.2
+
+  # Pad the stack because mem_load overwrites it
+  # stack: [PAD, read_ptr, write_ptr, count, words, ...]
+  padw
+
+  # LOOP: [PAD, read_ptr, write_ptr, count, words, ...]
+  # while(words>0) (17 cycles)
+  dup.7 dup.7 neq while.true
+    # perform read (2 cycles)
+    # stack: [A, read_ptr, write_ptr, count, words, ...]
+    dup.4 mem_loadw
+
+    # peform write (3 cycles)
+    # stack: [A, read_ptr, write_ptr, count, words, ...]
+    dup.5 mem_storew
+
+    # update counters (9 cycles)
+    # stack: [read_ptr+1, write_ptr+1, count+1, words, PAD, ...]
+    swapw
+        add.1 movup.3 movup.3 add.1 movup.3 add.1 movup.3
+    swapw 
+
+    dup.7 dup.7 neq # while(words>0) (3 cycles)
+  end
+
+  # clean stack
+  # stack: [...]
+  dropw dropw
+end
```

Comment:
Note: The original PR had a [comment](https://github.com/0xPolygonMiden/miden-base/pull/26#discussion_r1091541529) about using a counter with `add.1` instead of `sub.1` with the words. I wrote both versions, turns out it is the same number of cycles (unless I counted them wrong). `sub.1` is [`2` cycles](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/instruction/field_ops.rs#L39). The gain of `1` cycle from `add.1` goes away because of the stack manipulation of the counter.

Here are the two versions:

```
export.memcopy_with_counter
  # add counter
  # stack: [read_ptr, write_ptr, count, words, ...]
  push.0 movdn.2

  # Pad the stack because mem_load overwrites it
  # stack: [PAD, read_ptr, write_ptr, count, words, ...]
  padw

  # LOOP: [PAD, read_ptr, write_ptr, count, words, ...]
  # while(words>0) (17 cycles)
  dup.7 dup.7 neq while.true
    # perform read (2 cycles)
    # stack: [A, read_ptr, write_ptr, count, words, ...]
    dup.4 mem_loadw

    # peform write (3 cycles)
    # stack: [A, read_ptr, write_ptr, count, words, ...]
    dup.5 mem_storew

    # update counters (9 cycles)
    # stack: [read_ptr, write_ptr, words, count, PAD, ...]
    swapw
        add.1 movup.3 add.1 movup.3 movup.3 add.1 movup.3
    swapw 

    dup.7 dup.7 neq # while(words>0) (3 cycles)
  end

  # clean stack
  # stack: [...]
  dropw dropw
end
```

```
export.memcopy_without_counter
  # Pad the stack because mem_load overwrites it
  # stack: [PAD, read_ptr, write_ptr, words, ...]
  padw

  # LOOP: [PAD, read_ptr, write_ptr, words, ...]
  # while(words>0) (17 cycles per word)
  dup.6 neq.0 while.true
    # perform read (2 cycles)
    # stack: [A, read_ptr, write_ptr, words, ...]
    dup.4 mem_loadw

    # peform write (3 cycles)
    # stack: [A, read_ptr, write_ptr, words, ...]
    dup.5 mem_storew

    # update counters (9 cycles)
    swapw
        # stack: [read_ptr, write_ptr, words, x, PAD, ...]
        add.1 movup.2 sub.1 movup.2 add.1 movup.2
    swapw 

    dup.6 neq.0 # while(words>0) (3 cycles)
  end

  # clean stack
  # stack: [...]
  dropw drop drop drop
end
```

I'm not sure which one is the best, because with the counter and using `add.1` there is no need to encode immediate value, which I recently learned is a nice optimization. However the version without the counter uses fewer `3` fewer cycles overall. I suppose the perf difference is negligible, and we can bench it later on.

Reply:
> The original PR had a [comment](https://github.com/0xPolygonMiden/miden-base/pull/26#discussion_r1091541529) about using a counter with `add.1` instead of `sub.1`

We could also negate the counter before the loop and then still use `add.1` and terminate when we get to 0.